# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/credentialmanagement/identity/NavigatorLogin.cpp
# Commit: 0c5428075ca4
# Full Hash: 0c5428075ca4aa6b5270350247d43217fa5d0feb
# Author: Benjamin VanderSloot <bvandersloot@mozilla.com>
# Date: 2025-06-07 08:31:23
# Regressor Bug: 1969017
# File Overlap Count: 2
# Description:
#   Bug 1969017 - Refactor FedCM to use a managed actor rather than tagging onto WindowGlobalParent - r=anti-tracking-reviewers,emz
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D251617
# ==============================================================================

diff -r 9e7fe3dbea01 -r 0c5428075ca4 dom/credentialmanagement/identity/NavigatorLogin.cpp
--- a/dom/credentialmanagement/identity/NavigatorLogin.cpp	Fri Jun 06 14:38:54 2025 +0000
+++ b/dom/credentialmanagement/identity/NavigatorLogin.cpp	Fri Jun 06 14:38:54 2025 +0000
@@ -11,7 +11,7 @@
 #include "mozilla/dom/Promise.h"
 #include "mozilla/Maybe.h"
 #include "mozilla/net/SFVService.h"
-#include "mozilla/dom/WindowGlobalChild.h"
+#include "mozilla/dom/WebIdentityHandler.h"
 #include "nsCycleCollectionParticipant.h"
 #include "nsIGlobalObject.h"
 #include "nsIPermissionManager.h"
@@ -57,54 +57,30 @@
   return NavigatorLogin_Binding::Wrap(aCx, this, aGivenProto);
 }
 
-NavigatorLogin::NavigatorLogin(nsIGlobalObject* aGlobal) : mOwner(aGlobal) {
+NavigatorLogin::NavigatorLogin(nsPIDOMWindowInner* aGlobal) : mOwner(aGlobal) {
   MOZ_ASSERT(mOwner);
 };
 
 already_AddRefed<mozilla::dom::Promise> NavigatorLogin::SetStatus(
     LoginStatus aStatus, mozilla::ErrorResult& aRv) {
-  RefPtr<Promise> promise = Promise::Create(mOwner, aRv);
+  RefPtr<Promise> promise = Promise::Create(mOwner->AsGlobal(), aRv);
   if (aRv.Failed()) {
     return nullptr;
   }
 
-  nsPIDOMWindowInner* window = mOwner->GetAsInnerWindow();
-  if (!window) {
-    promise->MaybeRejectWithUnknownError(
-        "navigator.login.setStatus called on unavailable window"_ns);
-    return promise.forget();
-  }
-
-  if (!CredentialsContainer::IsSameOriginWithAncestors(window)) {
+  if (!CredentialsContainer::IsSameOriginWithAncestors(mOwner)) {
     promise->MaybeRejectWithSecurityError(
         "navigator.login.setStatus must be called in a frame that is same-origin with its ancestors"_ns);
     return promise.forget();
   }
 
-  WindowGlobalChild* wgc = window->GetWindowGlobalChild();
-  if (!wgc) {
-    promise->MaybeRejectWithUnknownError(
-        "navigator.login.setStatus called while window already destroyed"_ns);
+  WebIdentityHandler* identityHandler = mOwner->GetOrCreateWebIdentityHandler();
+  if (!identityHandler) {
+    promise->MaybeRejectWithOperationError("");
     return promise.forget();
   }
 
-  wgc->SendSetLoginStatus(aStatus)->Then(
-      GetCurrentSerialEventTarget(), __func__,
-      [promise](
-          const WindowGlobalChild::SetLoginStatusPromise::ResolveValueType&
-              aResult) {
-        if (NS_SUCCEEDED(aResult)) {
-          promise->MaybeResolveWithUndefined();
-        } else {
-          promise->MaybeRejectWithUnknownError(
-              "navigator.login.setStatus had an unexpected internal error");
-        }
-      },
-      [promise](const WindowGlobalChild::SetLoginStatusPromise::RejectValueType&
-                    aResult) {
-        promise->MaybeRejectWithUnknownError(
-            "navigator.login.setStatus had an unexpected internal error");
-      });
+  identityHandler->SetLoginStatus(aStatus, promise);
   return promise.forget();
 }
 